test/test_autograd.py

```diff
@@ -5450,6 +5450,56 @@ def test_advanced_indexing_backwards_large(self, device):
         a.sum().backward()
         self.assertEqual(x.grad, torch.ones(n, 1, device=device))
 
+    @onlyCUDA
+    def test_reentrant_parent_error_on_cpu(self, device):
+        before = CudaMemoryLeakCheck.get_cuda_memory_usage()
+        t1 = torch.rand([3, 3], requires_grad=True)
+        t2 = torch.rand([3, 3], device=device, requires_grad=True)
+        t3 = torch.rand([3, 3], device=device, requires_grad=True)
+
+        # Parent graph cpu graph.
+        t4 = t1 * t1
+        t5 = TestAutograd.BackwardError.apply(t4)
+
+        # Child gpu graph (much longer than parent graph).
+        computations = []
+        computations.append(t2 * t2)
+        for i in range(10):
+            computations.append(computations[-1] * t2)
+        reentrant_root = computations[-1].sum()
+
+        class ReentrantFunc(Function):
+            @staticmethod
+            def forward(ctx, inp):
+                return inp
+
+            @staticmethod
+            def backward(ctx, grad):
+                # Reentrant backward in child will take much longer.
+                reentrant_root.backward()
+                return grad
+
+        # Parent gpu graph.
+        t6 = ReentrantFunc.apply(t3)
+        t7 = t6 * t6
+
+        # Parent graph will error out first, while child graph will continue executing.
+        with self.assertRaisesRegex(RuntimeError, "simulate error"):
+            torch.autograd.backward([t5.sum(), t7.sum()])
+
+        # No grads should be accumulated since child graph will stop execution
+        # after parent receives error.
+        self.assertIsNone(t2.grad)
+        self.assertIsNone(t1.grad)
+        self.assertIsNone(t3.grad)
+
+        # Wait for autograd thread to cleanup failed tasks.
+        after = CudaMemoryLeakCheck.get_cuda_memory_usage()
+        start = time.time()
+        while before != after and time.time() - start < 30:
```

Comment:
The second condition is to avoid waiting forever here right?
Maybe raise an error when we reach it stating that we leaked something?

Reply:
When the unit test finishes the test framework already does this validation here: https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_utils.py#L538. So if we exceed that timeout without cleaning up all the memory, that check would fail and the unit test would fail as well.

Reply:
Ho ok, but you could raise a more informative error here like "Is the child graph stuck"?
If you don't do that, it would be nice to at least have a comment saying that the test infra is going to raise an error when we leak here.